housing_modeling/src/train.py

```diff
@@ -0,0 +1,148 @@
+
+housing_tr = pd.DataFrame(X, columns=housing_num.columns, index=housing.index)
+housing_tr["rooms_per_household"] = (
+    housing_tr["total_rooms"] / housing_tr["households"]
+)
+housing_tr["bedrooms_per_room"] = (
+    housing_tr["total_bedrooms"] / housing_tr["total_rooms"]
+)
+housing_tr["population_per_household"] = (
+    housing_tr["population"] / housing_tr["households"]
+)
+
+housing_cat = housing[["ocean_proximity"]]
+housing_prepared = housing_tr.join(
+    pd.get_dummies(housing_cat, drop_first=True)
+)
+
+lin_reg = LinearRegression()
+lin_reg.fit(housing_prepared, housing_labels)
+
+housing_predictions = lin_reg.predict(housing_prepared)
+lin_mse = mean_squared_error(housing_labels, housing_predictions)
+lin_rmse = np.sqrt(lin_mse)
+logger.info("linear regression rmse: %s" % (lin_rmse))
+lin_mae = mean_absolute_error(housing_labels, housing_predictions)
+logger.info("linear regression mae: %s" % (lin_mae))
+
+tree_reg = DecisionTreeRegressor(random_state=42)
+tree_reg.fit(housing_prepared, housing_labels)
+
+housing_predictions = tree_reg.predict(housing_prepared)
+tree_mse = mean_squared_error(housing_labels, housing_predictions)
+tree_rmse = np.sqrt(tree_mse)
+logger.info("Decision trees rmse: %s" % (tree_rmse))
+
+param_distribs = {
+    "n_estimators": randint(low=1, high=200),
+    "max_features": randint(low=1, high=8),
+}
+forest_reg = RandomForestRegressor(random_state=42)
+rnd_search = RandomizedSearchCV(
+    forest_reg,
+    param_distributions=param_distribs,
+    n_iter=10,
+    cv=5,
+    scoring="neg_mean_squared_error",
+    random_state=42,
+)
+rnd_search.fit(housing_prepared, housing_labels)
+cvres = rnd_search.cv_results_
+for mean_score, params in zip(cvres["mean_test_score"], cvres["params"]):
+    logger.info(
+        "random forest with random search gave score \n %s for parameters %s"
+        % (str(np.sqrt(-mean_score)), str(params))
+    )
+
+
+param_grid = [
+    # try 12 (3×4) combinations of hyperparameters
+    {"n_estimators": [3, 10, 30], "max_features": [2, 4, 6, 8]},
+    # then try 6 (2×3) combinations with bootstrap set as False
+    {"bootstrap": [False], "n_estimators": [3, 10], "max_features": [2, 3, 4]},
+]
+
+forest_reg = RandomForestRegressor(random_state=42)
+# train across 5 folds, that's a total of (12+6)*5=90 rounds of training
+grid_search = GridSearchCV(
+    forest_reg,
+    param_grid,
+    cv=5,
+    scoring="neg_mean_squared_error",
+    return_train_score=True,
+)
+grid_search.fit(housing_prepared, housing_labels)
+
+grid_search.best_params_
+cvres = grid_search.cv_results_
+for mean_score, params in zip(cvres["mean_test_score"], cvres["params"]):
+    logger.info(
+        "random forest with grid search gave score \n %s for parameters %s"
+        % (str(np.sqrt(-mean_score)), str(params))
+    )
+
+feature_importances = grid_search.best_estimator_.feature_importances_
+sorted(zip(feature_importances, housing_prepared.columns), reverse=True)
+
+
```

Comment:
Can you refactor this script into functions and add a `if __name__ == '__main__'` block to execute the functions.